README.rst
=====
pyPDA
=====


.. image:: https://img.shields.io/pypi/v/pypda.svg
        :target: https://pypi.python.org/pypi/pypda

.. image:: https://img.shields.io/travis/taoyilee/pypda.svg
        :target: https://travis-ci.com/taoyilee/pypda

.. image:: https://readthedocs.org/projects/pypda/badge/?version=latest
        :target: https://pypda.readthedocs.io/en/latest/?badge=latest
        :alt: Documentation Status




A Pythonic toolkit to generate synthetic blood pressure waveforms based on principle of pulse decomposition analysis (PDA).

* Free software: MIT license


Quick Start
-----------------


Use
~~~~~~~~~~~~~~~~
It is recommended that pyPDA should be setup in a Python `virtual environment <https://docs.python-guide.org/dev/virtualenvs/>`_.

.. code-block:: console

    python -m venv venv
    source venv/bin/activate
    pip install pyPDA
    pypda pulse --plot-dir work_dir --png # plot single beat waveform to word_dir

    # plot 10 second pulse waveform of 65 bpm, sampled at 90 Hz to word_dir
    pypda sample --bpm 65 --plot-dir work_dir --png --length 10 --sampling-rate 90

.. code-block:: python

    from pypda.pulse_model import PulseModelRaw

    pulse_waveform = PulseModelRaw()
    import matplotlib.pyplot as plt
    plt.plot(pulse_waveform.pulse_waveform)

    from pypda.wavelets import TriangGaussian
    triangular_pulse = TriangGaussian()
    triangular_pulse.shift(10) # optional horizontal shifting of waveform
    plt.plot(triangular_pulse.pulse_waveform)

    # addition operator is supported
    superpositioned_pulse = pulse_waveform + triangular_pulse
    plt.plot(superpositioned_pulse.pulse_waveform)


Development
~~~~~~~~~~~~~~~~~~~~~~~~
Please feel free to fork a copy on GitHUB. Pull requests are very welcome. To report an issue, please use `GitHub issue tracker <https://github.com/taoyilee/pyPDA/issues>`_.

.. code-block:: console

    python -m venv venv
    source venv/bin/activate
    git clone git@github.com:taoyilee/pyPDA.git
    pip install -r requirements_dev.txt
    # Start coding.

Screen Capture
---------------------
.. image:: docs/pulse.png
    :width: 400
    :alt: Sample blood pressure waveform

.. image:: docs/sample.png
    :width: 400
    :alt: Sample pulse waveform with abnormality indicator

References
---------------------
1. Baruch, Martin C., et al. "Pulse decomposition analysis of the digital arterial pulse during hemorrhage simulation." Nonlinear biomedical physics 5.1 (2011): 1. `[PDF] <https://www.ncbi.nlm.nih.gov/pmc/articles/PMC3025935/pdf/1753-4631-5-1.pdf>`_


HISTORY.rst
=======
History
=======

0.1.0 (2020-02-19)
------------------

* First release on PyPI.

0.2.1 (2020-03-02)
------------------

* Add sampler for test data (different parameters)


pypda/cli.py
"""Console script for pypda."""
import sys

import click

DEBUG = False


@click.group()
@click.option('--debug/--no-debug', default=False)
def cli(debug):
    global DEBUG
    DEBUG = debug
    click.echo('Debug mode is %s' % ('on' if debug else 'off'))


@cli.command()  # @cli, not @click!
@click.option('--plot-dir', default='plot', help='Plot output directory')
@click.option('--png/--no-png', default=False)
def wavelet(plot_dir='plot', png=False):
    import os
    os.makedirs(plot_dir, exist_ok=True)
    from pypda.wavelets import TriangGaussian
    import matplotlib.pyplot as plt
    w = TriangGaussian()
    w.shift(10)
    plt.plot(w.waveform)
    plt.grid()
    plt.title(f"Smoothed Triangular Window m_triangle={w.triang_m}, m_gaussian={w.gaussian_m} std={w.gaussian_std}")
    plt.ylabel("Amplitude")
    plt.xlabel("Sample")
    plt.tight_layout()
    if png:
        plt.savefig(os.path.join(plot_dir, "triangle.png"))


@cli.command()  # @cli, not @click!
@click.option('--plot-dir', default='plot', help='Plot output directory')
@click.option('--png/--no-png', default=False)
def pulse(plot_dir='plot', png=False):
    import os
    os.makedirs(plot_dir, exist_ok=True)
    from pypda.pulse_model import PulseModelRaw
    import matplotlib.pyplot as plt
    w = PulseModelRaw()
    plt.plot(w.waveform)
    for i, wavelet in enumerate(w.wavelets):
        plt.plot(wavelet.waveform, linestyle="--", label=f"component {i + 1}")
    plt.grid()
    plt.title(f"Pulse Model")
    plt.ylabel("Amplitude")
    plt.legend()
    plt.xlabel("Sample")
    plt.tight_layout()
    if png:
        plt.savefig(os.path.join(plot_dir, "pulse.png"))


@cli.command()  # @cli, not @click!
@click.option('--plot-dir', default='plot', help='Plot output directory')
@click.option('--length', default=10, help='Length of the sample in seconds')
@click.option('--bpm', default=65, help='BPM')
@click.option('--sampling-rate', default=90, help='Sampling rate Hz')
@click.option('--png/--no-png', default=False)
def sample(plot_dir='plot', length=10, bpm=65, sampling_rate=90, png=False):
    """

    :param plot_dir:
    :param length: in seconds
    :param sampling_rate: in Hz
    :param png:
    :return:
    """
    from pypda.pulse_wave import PulseWave
    import os
    os.makedirs(plot_dir, exist_ok=True)
    import matplotlib.pyplot as plt
    from pypda.parameters import PulseWaveParameters
    p = PulseWaveParameters()

    p.normal.beat_length.mean = 100
    p.abnormal.diastolic_baseline.mean = 90
    p.abnormal.diastolic_baseline.scale = 20
    p.abnormal.beat_length.mean = 0.6 * p.normal.beat_length.mean
    p.abnormal.pulse_amplitudes[1].mean = 3.5
    p.abnormal.pulse_amplitudes[2].mean = 4.5
    p.abnormal.pulse_amplitudes[3].mean = 2.2
    p.abnormal.pulse_amplitudes[4].mean = 1.1
    p.normal.scale = (
        10, 5, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2)

    p.abnormal.scale = (10, 20, .2, .2, .2, .2, .2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1)
    print(f"Parameters for normal =\n{p.normal}")
    print(f"Parameters for abnormal =\n{p.abnormal}")
    wave = PulseWave(parameters=p, bpm=bpm, length=length, sampling_rate=sampling_rate)
    x, y, t = wave.sample()

    plt.figure(figsize=(wave.beats, 6))
    plt.subplot(2, 1, 1)
    plt.title(f"Pulse Model")
    plt.ylabel("Blood Pressure (mmHg)")
    plt.xlabel("Time (Second)")
    plt.grid()
    plt.plot(t, x)
    plt.xlim([0, length])
    plt.subplot(2, 1, 2)
    plt.plot(t, y)
    plt.ylabel("Abnormality")
    plt.yticks([0, 1])
    plt.xlim([0, length])
    plt.tight_layout()
    if png:
        plt.savefig(os.path.join(plot_dir, "sample.png"))
    return x, y, t


@cli.command()  # @cli, not @click!
@click.option('--plot-dir', default='plot', help='Plot output directory')
@click.option('--length', default=10, help='Length of the sample in seconds')
@click.option('--bpm', default=65, help='BPM')
@click.option('--sampling-rate', default=90, help='Sampling rate Hz')
@click.option('--png/--no-png', default=False)
def sample_test(plot_dir='plot', length=10, bpm=65, sampling_rate=90, png=False):
    """

    :param plot_dir:
    :param length: in seconds
    :param sampling_rate: in Hz
    :param png:
    :return:
    """
    from pypda.pulse_wave import PulseWave
    import os
    os.makedirs(plot_dir, exist_ok=True)
    import matplotlib.pyplot as plt
    from pypda.parameters import PulseWaveParameters
    p = PulseWaveParameters()

    p.normal.beat_length.mean = 100
    p.abnormal.diastolic_baseline.mean = 86
    p.abnormal.diastolic_baseline.scale = 13
    p.abnormal.beat_length.mean = 0.3 * p.normal.beat_length.mean
    p.abnormal.pulse_amplitudes[1].mean = 3.1
    p.abnormal.pulse_amplitudes[2].mean = 4.1
    p.abnormal.pulse_amplitudes[3].mean = 2.1
    p.abnormal.pulse_amplitudes[4].mean = 0.9
    p.abnormal.delta_time[0].mean = 6
    p.abnormal.delta_time[1].mean = 7
    p.abnormal.delta_time[2].mean = 5
    p.abnormal.delta_time[3].mean = 7.3
    p.normal.scale = (
        10, 5, .2, .2, .2, .2, .2, .2, .2, .2, .2, .5, .5, .5, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2, .2)

    p.abnormal.scale = (10, 20, .2, .2, .2, .2, .2, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1)
    print(f"Parameters for normal =\n{p.normal}")
    print(f"Parameters for abnormal =\n{p.abnormal}")
    wave = PulseWave(parameters=p, bpm=bpm, length=length, sampling_rate=sampling_rate)
    x, y, t = wave.sample()

    plt.figure(figsize=(wave.beats, 6))
    plt.subplot(2, 1, 1)
    plt.title(f"Pulse Model")
    plt.ylabel("Blood Pressure (mmHg)")
    plt.xlabel("Time (Second)")
    plt.grid()
    plt.plot(t, x)
    plt.xlim([0, length])
    plt.subplot(2, 1, 2)
    plt.plot(t, y)
    plt.ylabel("Abnormality")
    plt.yticks([0, 1])
    plt.xlim([0, length])
    plt.tight_layout()
    if png:
        plt.savefig(os.path.join(plot_dir, "sample_test.png"))
    return x, y, t


if __name__ == "__main__":
    sys.exit(cli())  # pragma: no cover
